File: hashMap/49.cpp
#include<vector>
#include<string>
#include<algorithm>
#include<iostream>
#include<unordered_map>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        sort(strs.begin(), strs.end());
        unordered_map<string, int> my_map;
        int count = 0;
        vector<vector<string>> result;
        string element;
        for(int i=0; i<strs.size(); i++){
            element = strs[i];
            sort(element.begin(), element.end());
            if(my_map.find(element) == my_map.end()){
                my_map[element] = count;
                count++;
                vector<string> tmp;
                tmp.push_back(strs[i]);
                result.push_back(tmp);
            }else{
                result[my_map[element]].push_back(strs[i]);
            }
            
        }
        return result;
    }
};

int main(){
    string a[] = {"eat","tea","tan","ate","nat","bat"};
    vector<string> str(a, a+6);
    
    Solution s;
    vector<vector<string>> result=s.groupAnagrams(str);
    // cout<<result.size()<<endl;
    for(int i=0; i<result.size(); i++){
        for(int j=0; j<result[i].size(); j++){
            cout<<result[i][j]<<" ";
        }
        cout<<" "<<endl;
    }
    return -1;

}

File: hashMap/49.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2020-01-03 16:27:32
'''
import sys
import os

sys.path.append("..")

from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        text_dict = {}

        for s in strs:
            _s = "".join(sorted(s))
            if text_dict.get(_s) is None:
                text_dict[_s] = [s]
            else:
                text_dict[_s].append(s)
        
        
        result = [v for _, v in text_dict.items()]
        return result

if __name__ == "__main__":
    solu = Solution()

    print(solu.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))